File: LibraryApp/Application/Loans/LoanService.cs
using Application.Loans.DTOs;
using AutoMapper;
using Domain.Loans;

namespace Application.Loans;

public class LoanService : ILoanService
{
    private readonly ILoanRepository _loanRepository;
    private readonly IMapper _mapper;

    public LoanService(ILoanRepository loanRepository, IMapper mapper)
    {
        _loanRepository = loanRepository;
        _mapper = mapper;
    }

    public List<LoanDto> GetAll()
    {
        var loans = _loanRepository.GetAll();
        return _mapper.Map<List<LoanDto>>(loans);
    }

    public LoanDto GetById(int id)
    {
        var loan = _loanRepository.GetLoansByMemberId(id);
        return _mapper.Map<LoanDto>(loan);
    }
    
    public void Insert(LoanCreateDto vm)
    {
        var loanEntity = _mapper.Map<Loan>(vm);
        loanEntity.ReturnDate = null; 
        _loanRepository.Insert(loanEntity);
    }

    public void ReturnBook(int id)
    {
        _loanRepository.ReturnBook(id);
    }
    
    public List<LoanDto> GetLoansByMemberId(int id)
    {
        var loans = _loanRepository.GetLoansByMemberId(id);
        return _mapper.Map<List<LoanDto>>(loans);
    }
}

File: angular/src/app/features/loans/services/loan.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { LoanCreateDto, LoanDto } from '../models/loan.dto';
import { environment } from '../../../../environments/environment.development';

@Injectable({
    providedIn: 'root'
})
export class LoanService {
    private api_url = `${environment.apiUrl}/api/loans`;

    constructor(private http: HttpClient){}

    getAll(): Observable<LoanDto[]>{
        return this.http.get<LoanDto[]>(this.api_url);
    }

    getById(id: number): Observable<LoanDto>{
        return this.http.get<LoanDto>(`${this.api_url}/${id}`);
    }

    Insert(loan: LoanCreateDto): Observable<LoanDto>{
        return this.http.post<LoanDto>(this.api_url, loan);
    }

    getLoansByMemberId(memberId: number): Observable<LoanDto[]>{
        return this.http.get<LoanDto[]>(`${this.api_url}/member/${memberId}`);
    }

    returnBook(loanId: number): Observable<any>{
        return this.http.post<any>(`${this.api_url}/return/${loanId}`, {});
    }
}
